File: leetcode/c++/121_best_time_to_buy_and_sell_stock.cpp
// https://leetcode.cn/problems/best-time-to-buy-and-sell-stock/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min_cost = prices[0];
        int max_profit = 0;
        for (int price : prices) {
            min_cost = min(min_cost, price);
            max_profit = max(max_profit, price - min_cost);
        }
        return max_profit;
    }
};

// 时间复杂度为O(n)，空间复杂度为O(1)

File: leetcode/golang/121_best_time_to_buy_and_sell_stock.go
func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func maxProfit(prices []int) int {
    min_price := prices[0]
    max_profit := 0
    for i := 0; i < len(prices); i++ {
        min_price = min(min_price, prices[i])
        max_profit = max(max_profit, prices[i] - min_price)
    }
    return max_profit
}